[wip] Introduce documentation about managing a separate mount namespace (#5724)
kubernetes/community: contributors/devel/sig-node/separate-mount-namespace.md

-t 1 -m` (assuming they have the host OS's /proc filesystem mounted so they can
reach the mount namespace of the real init process in /proc/1/ns/mnt)

### Legacy Tools

Any monitoring tool that relies on the ability to run in the host OS and have
visibility of mountpoints created by Kubelet, the Container Runtime, or
containers themselves, will need to enter the container mount namespace in
order to see these mountpoints.  This could be done by:

- Altering the legacy tool to be cloud-native
- Altering the scripts that launch the legacy tool to run it inside the
  container-mount namespace
- Altering the legacy tool to be aware of the container-mount namespace and
  enter it as needed to probe or interact with the container-specific
  mountpoints
- Providing a system-wide mechanism to revert the Kubernetes installation back
  to the original "all mounts in the Host OS" mode of operation.

### Administrative Access

An administrator logged in to the host OS wishing to obtain information about
the Kubernetes system for debugging or auditing purposes will need to be aware
of the new mount namespace, and will need to run 'nsenter' to view or interact
with these mountpoints.

> ## Implementation Suggestions

Reviewer: @lack Let's add an explicit note that there are no changes proposed or needed to kubelet or other components to use this new style of implementation.

visibility of mountpoints created by Kubelet, the Container Runtime, or
containers themselves, will need to enter the container mount namespace in
order to see these mountpoints.  This could be done by:

- Converting the legacy tool to be containerized
- Altering the scripts that launch the legacy tool to run it inside the
  container-mount namespace
- Altering the legacy tool to be aware of the container-mount namespace and
  enter it as needed to probe or interact with the container-specific
  mountpoints
- Providing a system-wide mechanism to revert the Kubernetes installation back
  to the original "all mounts in the Host OS" mode of operation.

### Administrative Access

An administrator logged in to the host OS wishing to obtain information about
the Kubernetes system for debugging or auditing purposes will need to be aware
of the new mount namespace, and will need to run `nsenter` to view or interact
with these mountpoints.

## Implementation Suggestions

This does not require a change to Kubelet, the Container Runtime, or any other
existing component, but can be accomplished entirely through changes to the

> environment and init scripts which launch Kublete and the Container Runtime.

Reviewer 2: would it be desirable to have end-to-end test configured this way so we know that this suggestion actually works and no new changes will break it? How hard would it be?

Author: Yes, that's a good idea.  I'd need to learn more about what existing environments there are that run the e2e tests before I could tell you how hard it would be; Do you have any pointers to where these environments are defined or set up?

Reviewer 2: @ike-ma spent some time recently thinking how to enable swap for tests. Ike, do you have quick pointers on what's the bet way to customize e2e tests for the scenario in this PR?

# Managing a Separate Mount Namespace for Kubernetes

## What does this mean?

Originally, Kubelet and the Container Runtime all ran in the same mount namespace as
the host OS (systemd, login shells, system daemons).

![Fig 1: Original mount propagation](images/original-k8s-mount-propagation.png)

_Fig 1: Original mount propagation_

However, in order to support the full range of cloud-native applications, only
two guarantees need to be made about the relationship of mountpoints:

1. Mounts originating in the host OS are visible to the Container Runtime, OCI
   hooks, Kubelet, and container volumeMounts with `MountPropagation:
   HostToContainer` (or `Bidirectional`)
2. Mounts originating in the Container Runtime, OCI hooks, Kubelet, and
   container volumeMounts with `mountPropagation: Bidirectional` are visible to
   each other and container volumeMounts with `MountPropagation:
   HostToContainer`

The fact that mounts originating in the Container Runtime, OCI hooks, Kubelet,
and container volumeMounts with `mountPropagation: Bidirectional` were visible

> to the host OS in the original implementations of Kubernetes was accidental to

Reviewer 3: ```suggestion
to the host OS in the original implementations of Kubernetes was incidental to
```

the implementation, and there are a number of real-world benefits to changing
the way Kubernetes is installed in order to hide the container-specific mount
points from the host OS.

![Fig 2: Container-Runtime-scoped mount propagation](images/hidden-k8s-mount-propagation.png)

_Fig 2: Container-Runtime-scoped mount propagation_

## Benefits

### Systemd Efficiency

Today systemd's CPU usage is adversely affected by a large number of

> mountpoints, especially versions released prior to December 2020. This feature

Reviewer 3: Is there a specific semantic version we could include here?

Author: Good idea.

Author: Found it.  `systemd-248`  Updated with actual release version.

of systemd, which scans all /proc/1/mount to invoke udev events for every
change suffers from two issues. The first was an inefficiency in systemd’s
processing logic. This was fixed in December 2020, and this fix reportedly
reduced the systemd mount-scanning-related overhead by 30-50%. The second is a
lack of granularity in the events the kernel sends to systemd about mountpoint
changes, which necessitates a full rescan of all mountpoints even on a minor
change or single mountpoint addition. The kernel change to remedy this seems
unlikely to move forward in the near term.  (See
[BZ1819869](https://bugzilla.redhat.com/show_bug.cgi?id=1819868) for more

> details)

Reviewer 3: This seems RHEL/Red Hat specific so I would drop it.

For example, a fairly small OpenShift deployment today (which does not yet have
the systemd December 2020 fix) has upwards of 400 mountpoints on every node
owned by Kubelet, the Container Runtime, and infrastructure containers like OVN

> or CSI operators. Hiding the container mountpoints reduces the steady-state

Reviewer 3: Do we have an upstream Kubernetes parallel use case we could reference here instead?

Author: A great question.  I don't have one, but it may be worth setting on up.

systemd CPU utilization from ~10% to ~0% on every worker node. On a more
heavily loaded system, with 1000’s of container mountpoints, systemd’s usage
will cap out at 100% of a single CPU core. Even with the 30-50% reduction in
systemd usage expected from the December 2020 fix, having systemd take up as
much as 50-70% of a single CPU in the worst-case scenario is a lot of
unnecessary overhead.

Administrators of other OSes with other host OS services that monitor or
process mountpoints indiscriminately could also benefit from hiding these
container-specific mountpoints from the OS at large.

### Tidiness and Encapsulation

These container overlay mounts, secret and configmap mounts, and namespace
mounts are all internal implementation details around the way the Kubernetes
and the Container Runtime allocate resources to containers. For normal
day-to-day operation, there is no need for these special-use mounts to be

> visible in the host OS's top-level 'mount' or 'findmnt' output.

Reviewer 3: ```suggestion
visible in the host OS's top-level `mount` or `findmnt` output.
```

> ### Security

Reviewer 3: cc @kubernetes/sig-security-pr-reviews

Author: Hey, turns out this proposed mechanism likely prevents containers (even privileged containers) from exploiting [CVE-2021-33910](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-33910)!

While access to container-specific mounts in Kubernetes (like secrets) today
are protected by filesystem permissions, the fact that the locations are

> visible to unprivileged users via the 'mount' command can provide information

Reviewer 3: ```suggestion
visible to unprivileged users via the `mount` command can provide information
```

which could be used in conjunction with other vulnerabilities to decipher
secrets or other protected container mount contents. Locking access to even
listing the mountpoints behind the ability to enter the mount namespace (which
in turn requires `CAP_SYS_CHROOT` and `CAP_SYS_ADMIN`) removes one level of
potential access to container-specific mounts.

### No one uses it (that we know of)

After much investigation and conversation on both the sig-node and sig-storage
mailing lists, as well as within Red Hat's OpenShift communities, we have yet
to turn up a use case that relies on this container-to-host propagation.

### No one _should_ use it

In fact, it may be seen as an anti-pattern and run contrary to the
container-native philosophy. Relying on a host OS to interact with a

> container-owned mountpoint seems suspect.

Reviewer 3: It's not clear to me from this paragraph why this is an anti-pattern. I'd suggest we either drop this paragraph or substantiate this a bit better :)

Author: Sure, never hurts to try to clarify my assumptions here...

So I was basically thinking that the cloud-native way of doing things is to make sure an end-user's full solution works by being distributed in a set of containers.  It seems obvious to me that requiring a special package on the Host OS to interact with something your containers do is not cloud-native.  Do you agree?  Can you think of a better way I could say that?

Reviewer 3: I don't think this is obvious, and I'd shy away from making universal recommendations of good or bad, because there's always an exception to break the rule... I'd just remove this section, I think :)

### Testing implications

With separate mount namespaces, it would be possible to run multiple instances
of Kubernetes on a single host without the various mountpoints interfering with
one another. There may be other inter-instance difficulties besides the mount
namespace, but this does move in that direction without resorting to more
complicated solutions like trying to run Kubernetes inside of containers.

## Costs and Implications

> If a Kubernetes distribution intends to implement this mount-hiding mode of

Reviewer 3: ```suggestion
If a Kubernetes deployment intends to implement this mount-hiding mode of
```

operation, there are a number of costs and implications that should be
documented for system administrators.

### Privileged pods

Privileged pods that previously expected to have direct access to the host's

> root mount namespace (such as the one spawned by `oc debug`), will now be

Reviewer 3: Do we have a kubectl specific use case here we could use?

Author: Not that I know of, but I'm open to suggestions.

Reviewer 3: @SergeyKanzhelev is there an upstream example we could use here?

rooted in the new container-runtime mount namespace.  They will still have
visibility into the host OS's mount points as granted by the normal volumeMount
permissions, but any mounts made in these privileged pods will not by default
propagate up to be visible by the Host OS. In order for shells or utilities in
privileged pods like this to propagate mounts up to the host OS, they can enter
the host OS namespace by running `nsenter -t 1 -m` (assuming they have the host
OS's /proc filesystem mounted so they can reach the mount namespace of the real
init process in /proc/1/ns/mnt)

### Legacy Tools

Any monitoring tool that relies on the ability to run in the host OS and have
visibility of mountpoints created by Kubelet, the Container Runtime, or
containers themselves, will need to enter the container mount namespace in
order to see these mountpoints.  This could be done by:

> - Altering the legacy tool to be cloud-native

Reviewer 3: What does this mean?

Author: "Put it in a container", but I agree this can be clearer.

- Altering the scripts that launch the legacy tool to run it inside the
  container-mount namespace
- Altering the legacy tool to be aware of the container-mount namespace and
  enter it as needed to probe or interact with the container-specific
  mountpoints
- Providing a system-wide mechanism to revert the Kubernetes installation back
  to the original "all mounts in the Host OS" mode of operation.

### Administrative Access

An administrator logged in to the host OS wishing to obtain information about
the Kubernetes system for debugging or auditing purposes will need to be aware

> of the new mount namespace, and will need to run 'nsenter' to view or interact

Reviewer 3: ```suggestion
of the new mount namespace, and will need to run `nsenter` to view or interact
```

with these mountpoints.

## Implementation Suggestions

This does not require a change to Kubelet, the Container Runtime, or any other
existing component, but can be accomplished entirely through changes to the
environment and init scripts which launch Kublete and the Container Runtime.

### Mount Namespace Creation

The new mount namespace must be a `slave+shared` mountpoint (according to
[mount_namespace(7)](https://man7.org/linux/man-pages/man7/mount_namespaces.7.html))
descended from the top-level mount namespace.

> - It needs to be a `slave` to the host OS namespace so it sees the host

Reviewer 3: Unfortunately this terminology has not been updated upstream so it's required for accuracy in these documents... not sure what the process is with the Inclusive Naming Initiative (INI) to address this sort of thing.

Author: I know.  I suspect there will be a lot of historical baggage to clear out that will take some time to rework.

Author: I was actually hoping the reference to the namespace above would be a bit of a "Sorry we couldn't change that on our own" pointer.

Author: I asked on the INI mailing list, and I think the most reasonable suggestion I saw there was to refer to the legacy term once and use preferred terms thereafter, except as needed for technical accuracy.  I'll re-read this through and see if I can do better...

Author: Yes, I have removed one instance of "slave", and I think it's still clear and accurate.
